src/token-processor/blockchain-api/blockchain-smart-contract-monitor.ts

```diff
@@ -0,0 +1,126 @@
+import * as postgres from 'postgres';
+import { PgStore } from '../../pg/pg-store';
+import { PgBlockchainApiStore } from '../../pg/blockchain-api/pg-blockchain-api-store';
+import {
+  getContractLogMetadataUpdateNotification,
+  getSmartContractSip,
+} from '../util/sip-validation';
+import { ClarityAbi } from '@stacks/transactions';
+import { Static, Type } from '@sinclair/typebox';
+import { TypeCompiler } from '@sinclair/typebox/compiler';
+import { logger } from '../../logger';
+
+const PgNotification = Type.Object({
+  type: Type.String(),
+  payload: Type.Object({}, { additionalProperties: true }),
+});
+const PgNotificationCType = TypeCompiler.Compile(PgNotification);
+
+const PgSmartContractPayload = Type.Object({ contractId: Type.String() });
+const PgSmartContractPayloadCType = TypeCompiler.Compile(PgSmartContractPayload);
+type PgSmartContractPayloadType = Static<typeof PgSmartContractPayload>;
+
+const PgSmartContractLogPayload = Type.Object({ txId: Type.String(), eventIndex: Type.Integer() });
+const PgSmartContractPayloadLogCType = TypeCompiler.Compile(PgSmartContractLogPayload);
+type PgSmartContractPayloadLogType = Static<typeof PgSmartContractLogPayload>;
+
+/**
+ * Listens for postgres notifications emitted from the API database when new contracts are deployed
+ * or contract logs are registered. It will analyze each of them to determine if they're new token
+ * contracts that need indexing or SIP-019 notifications that call for a token metadata refresh.
+ */
+export class BlockchainSmartContractMonitor {
+  private readonly db: PgStore;
+  private readonly apiDb: PgBlockchainApiStore;
+  private listener?: postgres.ListenMeta;
+
+  constructor(args: { db: PgStore; apiDb: PgBlockchainApiStore }) {
+    this.db = args.db;
+    this.apiDb = args.apiDb;
+  }
+
+  async start() {
+    try {
+      this.listener = await this.apiDb.sql.listen(
+        'stacks-api-pg-notifier',
+        message => void this.handleMessage(message),
+        () => logger.info(`BlockchainSmartContractMonitor connected`)
+      );
+    } catch (error) {
+      logger.error(`BlockchainSmartContractMonitor unable to connect`, error);
+      throw error;
+    }
+  }
+
+  async stop() {
+    await this.listener
+      ?.unlisten()
+      .then(() => logger.info(`BlockchainSmartContractMonitor connection closed`));
+  }
+
+  protected async handleMessage(message: string) {
+    const messageJson = JSON.parse(message);
+    if (PgNotificationCType.Check(messageJson)) {
```

Comment:
Seems like the majority of notifications are not relevant for this service, thoughts on starting to separate different notification payloads types into different channels in order to reduce overhead? I'm not sure what effect this has in production.

Reply:
Good idea... Opened an API issue to keep track of this https://github.com/hirosystems/stacks-blockchain-api/issues/1508